Feature/add struct array datatype to wrangler

sechegaray opened the pull request.

tivv commented on wrangler-core/src/main/java/io/cdap/directives/row/Flatten.java:
```diff
@@ -110,6 +110,11 @@ public void destroy() {
       }
 
       if (max == 0) {
+        for (int i = 0; i < count; ++i) {
```
I believe this is part of other PR

tivv commented on wrangler-core/src/main/java/io/cdap/wrangler/utils/RecordConvertor.java:
```diff
@@ -69,6 +69,9 @@
    * @return Populated {@link StructuredRecord}
    */
   public StructuredRecord decodeRecord(Row row, Schema schema) throws RecordConvertorException {
+    if (row == null) {
```
Please annotate parameter and return value as `@Nullable`

tivv commented on wrangler-core/src/main/java/io/cdap/wrangler/utils/RecordConvertor.java:
```diff
@@ -87,7 +90,7 @@ public StructuredRecord decodeRecord(Row row, Schema schema) throws RecordConver
       Object value = null;
       int idx = -1;
-      if (name.equals(row.getColumn(firstUnclaimedField))) {
+      if ((firstUnclaimedField < row.width()) && (name.equals(row.getColumn(firstUnclaimedField)))) {
```
Could you please make this a separate PR and cherry-pick into current release?

tivv commented on wrangler-core/src/main/java/io/cdap/wrangler/utils/RecordConvertor.java:
```diff
@@ -196,8 +199,14 @@ private Object decode(String name, Object object, Schema schema) throws RecordCo
   }
 
   private StructuredRecord decodeRecord(String name, Object object, Schema schema) throws RecordConvertorException {
+    if (object == null) {
```
Same here. If you need to treat something as nullable, please annotate it

tivv commented on wrangler-core/src/main/java/io/cdap/wrangler/utils/SchemaConverter.java:
```diff
@@ -152,6 +155,16 @@ public Schema getSchema(Object value, String name) throws RecordConvertorExcepti
+          fields.add(Field.of(field.getFirst(), getSchema(field.getSecond(), field.getFirst())));
+        }
+      }
+      return Schema.nullableOf(Schema.recordOf(UUID.randomUUID().toString().replace("-", ""), fields));
```
Please use stable names for schema as discussed

tivv commented on wrangler-core/src/main/java/io/cdap/wrangler/utils/SchemaConverter.java:
```diff
@@ -225,7 +238,7 @@ private Schema toSchema(String name, JsonObject object) throws RecordConvertorEx
       return Schema.nullableOf(Schema.arrayOf(Schema.of(Schema.Type.NULL)));
     }
-    return Schema.recordOf(name, fields);
+    return Schema.recordOf(UUID.randomUUID().toString().replace("-", ""), fields);
```
Please use stable names as discussed, also make it backwards-compatible

tivv commented on wrangler-core/src/test/java/io/cdap/directives/row/FlattenTest.java:
```diff
@@ -45,7 +45,7 @@ public void testEmptyList() throws Exception {
     Assert.assertEquals(1, output.size());
     Assert.assertEquals("y", output.get(0).getValue("y"));
-    Assert.assertEquals(0, ((List) output.get(0).getValue("x")).size());
+    Assert.assertNull(((List) output.get(0).getValue("x")));
```
I believe this is part of other PR

tivv commented on wrangler-service/src/main/java/io/cdap/wrangler/service/directive/AbstractDirectiveHandler.java:
```diff
@@ -97,6 +107,76 @@ public void destroy() {
+          if (value == null) {
+            return;
+          }
+          final TypeAdapter<Object> elementAdapter = gson.getAdapter(Object.class);
```
nit: move this into `rowClassAdapter` method. This would ensure it's called once and not for each record, similar to `defaultClassAdapter`

tivv commented on wrangler-service/src/main/java/io/cdap/wrangler/service/directive/AbstractDirectiveHandler.java:
```diff
@@ -97,6 +107,76 @@ public void destroy() {
+    }
+
+    private TypeAdapter<Object> defaultClassAdapter(Gson gson) {
+      final TypeAdapter<Object> elementAdapter = gson.getAdapter(Object.class);
```
is this used at all?

tivv commented on wrangler-service/src/main/java/io/cdap/wrangler/service/directive/AbstractDirectiveHandler.java:
```diff
@@ -97,6 +107,76 @@ public void destroy() {
+            if ((value.getClass().getMethod("toString").getDeclaringClass() != Object.class)) {
+              out.value(value.toString());
+            } else {
+              out.value("Non-displayable object");
```
Could you please confirm with product on what to say here. I'd prefer something shorter as those JSONs going to be displayed as 1 line in UI

tivv commented on wrangler-service/src/main/java/io/cdap/wrangler/service/directive/AbstractDirectiveHandler.java:
```diff
@@ -146,6 +238,7 @@ protected DirectiveExecutionResponse generateExecutionResponse(
     return new DirectiveExecutionResponse(values, headers, types, getWorkspaceSummary(rows));
   }
 
+
```
Do you need this extra empty line?

tivv reviewed (commented):
Please add a feature flag to ensure we don't expose this feature until after it's fully implemented and tested

yaojiefeng commented on wrangler-core/src/main/java/io/cdap/directives/column/CreateRecord.java:
```diff
@@ -0,0 +1,110 @@
+import java.util.List;
+
+/**
+ * A directive that Creates a record from columns
```
Creates -> creates

yaojiefeng commented on wrangler-core/src/main/java/io/cdap/directives/column/ExplodeRecord.java:
```diff
@@ -0,0 +1,123 @@
+  private String[] columns;
+
+  // Map to Keep Track new Columns;
+  private ArrayList<String> newColumnNames = new ArrayList<>();
```
`List<String>`, no need to use concrete object type

yaojiefeng replied:
This object seems only used to track the generated columns for lineage purpose, and is updated in the execute method. This won't work since the `lineage()` call is invoked in `prepareRun`, which is before the `execute` call

yaojiefeng commented on wrangler-core/src/main/java/io/cdap/directives/column/ExplodeRecord.java:
```diff
@@ -0,0 +1,123 @@
+
+          if (!(object instanceof Row)) {
+            throw new DirectiveExecutionException(
+                NAME,
```
nit - indentation should be 2 characters instead of 4.

yaojiefeng commented on wrangler-core/src/main/java/io/cdap/directives/column/ExplodeRecord.java:
```diff
@@ -0,0 +1,123 @@
+  @Override
+  public Mutation lineage() {
+    return Mutation.builder()
+        .readable("Flatten the record in columns %s into columns", Arrays.asList(columns))
```
nit - indentation should be 2 characters instead of 4.

yaojiefeng commented on wrangler-core/src/main/java/io/cdap/directives/column/ExplodeRecord.java:
```diff
@@ -0,0 +1,123 @@
+    return Mutation.builder()
+        .readable("Flatten the record in columns %s into columns", Arrays.asList(columns))
+        .all(Many.columns(columns), Many.columns(newColumnNames.toArray(new String[0]))
+        )
```
this doesn't need an extra line

yaojiefeng commented on wrangler-core/src/main/java/io/cdap/directives/column/ExplodeRecord.java:
```diff
@@ -0,0 +1,123 @@
+    for (Row row : rows) {
+      for (String column: columns) {
+        int idx = row.find(column);
+        if (idx != -1) {
```
better change it to 
```
if (idx == -1) {
  results.add(row); 
  continue;
}
...
```
to avoid the reduce the nested lines

sechegaray replied:
done

yaojiefeng commented on wrangler-core/src/main/java/io/cdap/wrangler/utils/RecordConvertor.java:
```diff
@@ -64,11 +65,15 @@
-  public StructuredRecord decodeRecord(Row row, Schema schema) throws RecordConvertorException {
+  @Nullable
+  public StructuredRecord decodeRecord(@Nullable Row row, Schema schema) throws RecordConvertorException {
+    if (row == null) {
```
Under what condition this will be null? I feel in Wrangler, the row will not be null right?

sechegaray replied:
A Nested Record, is converted to A Row inside a Row in Wrangler.  It can be that the nested record is nullable.

yaojiefeng commented on wrangler-core/src/main/java/io/cdap/directives/column/ExplodeRecord.java:
```diff
@@ -0,0 +1,123 @@
+                    column, object.getClass().getSimpleName()));
+          }
+
+          Row rowColumns = ((Row) object);
```
((Row) object) -> (Row) object

yaojiefeng commented on wrangler-core/src/main/java/io/cdap/directives/column/ExplodeRecord.java:
```diff
@@ -0,0 +1,123 @@
+public class ExplodeRecord implements Directive, Lineage {
+  public static final String NAME = "explode-record";
+  // Column on which to apply mask.
+  private String[] columns;
```
I don't understand why this directive has to work on a list of columns, shouldn't it just operate on one column, and if a user wants to explode records on multiple columns, he can invoke this directive multiple times on different columns. Making this a list might result in a large change in the workspace which might be confusing to user

sechegaray replied:
There's already a directive called `Flatten` that works for arrays. That directive takes a list of columns. Making a different syntax for two similar directives might confuse the users even more.

yaojiefeng replied:
I still think this should just operate on a single column, it doesn't make sense for this to operate on multiple columns, otherwise all the directives related to single column should be extended to be multiple columns. We shouldn't use old wrong behavior as tradition. With new implementation, we should make sure things are correct

bdmogal replied:
I think we should actually support multiple columns. The key is that we don't support many transformations on nested data. You *have* flatten them. So providing an option to do it in a single directive does seem more user friendly.

yaojiefeng commented on wrangler-core/src/main/java/io/cdap/directives/column/ExplodeRecord.java:
```diff
@@ -0,0 +1,123 @@
+          for (Pair<String, Object> field : rowColumns.getFields()) {
+            columnName = String.format("%s_%s", column, field.getFirst());
+            newColumnNames.add(columnName);
+            row.add(columnName, field.getSecond());
```
Should we use `addOrSet` here? If there is a conflict in the column name, this change will result in two columns with same name, which can cause weird behavior in the following directives execution

sechegaray replied:
Good Point

yaojiefeng commented on wrangler-core/src/main/java/io/cdap/directives/column/ExplodeRecord.java:
```diff
@@ -0,0 +1,123 @@
+import java.util.List;
+
+/**
+ * A directive for splitting a RowType ("Record") Column into multiple Columns.
```
We need much better documentation here, especially on what the resulting column names will look like(exploded col + "_" + original col name) will be after flatten, it should also mention what will happen if there is a conflict in the colume name after flatten, whether the existing column is replaced or a column with same name will be added(which should not be the case since we don't allow field with same name in StructuredRecord)

yaojiefeng commented on wrangler-core/src/main/java/io/cdap/directives/column/ExplodeRecord.java:
```diff
@@ -0,0 +1,123 @@
+  public Mutation lineage() {
+    return Mutation.builder()
+        .readable("Flatten the record in columns %s into columns", Arrays.asList(columns))
+        .all(Many.columns(columns), Many.columns(newColumnNames.toArray(new String[0]))
```
Our lineage does not work well for nested fields, I think it is fine if this directive does not extend from `Lineage`

sechegaray replied:
Talked with @bdmogal about this. We will document it as a known issue. The same issue occurs with other directives that generate new columns dynamically.

yaojiefeng commented on wrangler-core/src/main/java/io/cdap/wrangler/utils/SchemaConverter.java:
```diff
@@ -152,6 +168,23 @@ public Schema getSchema(Object value, String name) throws RecordConvertorExcepti
     }
 
+    if (value instanceof Row) {
+      ArrayList<Field> fields = new ArrayList<>();
```
`List<Field>`

yaojiefeng commented on wrangler-core/src/main/java/io/cdap/wrangler/utils/SchemaConverter.java:
```diff
@@ -208,13 +245,25 @@ private Schema toSchema(String name, JsonPrimitive primitive) throws RecordConve
+  }
+
+  private Schema toSchema(String name,
+                        JsonObject object, String recordPrefix) throws RecordConvertorException {
```
nit - indentation off

yaojiefeng replied:
Mark `recordPrefix` as `@Nullable` if it can be null.

yaojiefeng commented on wrangler-core/src/main/java/io/cdap/wrangler/utils/SchemaConverter.java:
```diff
@@ -81,14 +84,27 @@ public Schema toSchema(String id, Row row) throws RecordConvertorException {
+   * @return the schema of this object
+   */
+  @Nullable
+  public Schema getSchema(Object value, String name, String recordPrefix) throws RecordConvertorException {
```
Mark `recordPrefix` as `@Nullable` if it can be null.

yaojiefeng commented on wrangler-core/src/main/java/io/cdap/wrangler/utils/StructuredToRowTransformer.java:
```diff
@@ -77,14 +78,36 @@ public static Object getValue(StructuredRecord input, String fieldName) {
+        return val instanceof StructuredRecord ? transform((StructuredRecord) val) : val;
+      case ARRAY:
+        if (val instanceof Iterable) {
+          ArrayList<Object> rowList = new ArrayList<>();
```
`List<Object>`

yaojiefeng commented on wrangler-core/src/main/java/io/cdap/wrangler/utils/StructuredToRowTransformer.java:
```diff
@@ -77,14 +78,36 @@ public static Object getValue(StructuredRecord input, String fieldName) {
+        return val instanceof ByteBuffer ? Bytes.toBytes((ByteBuffer) val) : val;
+      // Recursively process structured records.
+      case RECORD:
+        return val instanceof StructuredRecord ? transform((StructuredRecord) val) : val;
```
what other type can `val` have here? Should we throw exception if it is not a StructuredRecord?

yaojiefeng commented on wrangler-core/src/main/java/io/cdap/wrangler/utils/StructuredToRowTransformer.java:
```diff
@@ -77,14 +78,36 @@ public static Object getValue(StructuredRecord input, String fieldName) {
+      case RECORD:
+        return val instanceof StructuredRecord ? transform((StructuredRecord) val) : val;
+      case ARRAY:
+        if (val instanceof Iterable) {
```
what other type can val have here? Should we throw exception if it is not a `Iterable`?

sechegaray replied:
Added the exception.

yaojiefeng commented on wrangler-core/src/test/java/io/cdap/wrangler/utils/StructuredToRowTransformerTest.java:
```diff
@@ -53,10 +56,20 @@ public void testStructuredRecordToRow() {
       Schema.Field.of("f12", Schema.of(Schema.LogicalType.TIME_MILLIS)),
       Schema.Field.of("f13", Schema.decimalOf(3, 2)),
-      Schema.Field.of("f14", Schema.of(Schema.LogicalType.DATETIME)));
+      Schema.Field.of("f14", Schema.of(Schema.LogicalType.DATETIME))
```
Should also add `Array` type to the record for unit test.

sechegaray replied:
Added

yaojiefeng commented on wrangler-service/src/main/java/io/cdap/wrangler/service/directive/AbstractDirectiveHandler.java:
```diff
@@ -146,6 +159,76 @@ public void destroy() {
+
+        @Override public
+        Row read(JsonReader in) throws IOException {
+          throw new UnsupportedOperationException("Reading Rows from Wrangler display format not implemented");
```
`not implemented` -> `not supported`

yaojiefeng commented on wrangler-service/src/main/java/io/cdap/wrangler/service/directive/AbstractDirectiveHandler.java:
```diff
@@ -176,16 +262,25 @@ protected DirectiveExecutionResponse generateExecutionResponse(
             schema = schema.isNullable() ? schema.getNonNullable() : schema;
-            type = schema.getLogicalType() == null ? schema.getType().name() : schema.getLogicalType().name();
+            type = schema.getLogicalType() == null ? schema.getType().name()
+                : schema.getLogicalType().name();
```
this doesn't need a new line

yaojiefeng commented on wrangler-service/src/main/java/io/cdap/wrangler/service/directive/AbstractDirectiveHandler.java:
```diff
@@ -146,6 +159,76 @@ public void destroy() {
+   * complex types by callint toString().  This is done so the complex types are shown
+   * in a similar way inside and outside the structure records.
+   */
+  private static class WranglerDisplaySerializer implements TypeAdapterFactory {
```
better separate this to a new class and have unit tests for it

yaojiefeng reviewed (commented):
Took a first pass, have some comments

rmstar commented on wrangler-core/src/main/java/io/cdap/directives/column/CreateRecord.java:
```diff
@@ -0,0 +1,110 @@
+      Row newRecord = new Row();
+      for (String columnName : columns) {
+        Object columnValue = row.getValue(columnName);
+        // Only set value y column is set.
```
y -> if ?

sechegaray replied:
changed

rmstar commented on wrangler-core/src/main/java/io/cdap/directives/column/CreateRecord.java:
```diff
@@ -0,0 +1,110 @@
+    List<String> cols = ((ColumnNameList) args.value("columns")).value();
+    targetColumn = args.value("target_column").value().toString();
+    columns = new String[cols.size()];
+    columns = cols.toArray(columns);
```
nit: can be simplified - columns = cols.toArray(new String[0])

rmstar commented on wrangler-service/src/main/java/io/cdap/wrangler/service/directive/AbstractDirectiveHandler.java:
```diff
@@ -156,6 +239,9 @@ protected DirectiveExecutionResponse generateExecutionResponse(
     SchemaConverter convertor = new SchemaConverter();
 
+
+    Gson gson = new GsonBuilder().registerTypeAdapterFactory(new WranglerDisplaySerializer()).create();
```
Can we create a static Gson instance instead of creating it here?

yaojiefeng commented on wrangler-core/src/main/java/io/cdap/directives/column/ExplodeRecord.java:
```diff
@@ -0,0 +1,141 @@
+  private String[] columns;
+
+  // Map to Keep Track new Columns;
+  private List<String> newColumnNames = new ArrayList<>();
```
Can just remove this if we do not care about lineage

yaojiefeng commented on wrangler-core/src/main/java/io/cdap/directives/column/ExplodeRecord.java:
```diff
@@ -0,0 +1,141 @@
+  }
+
+  @Override
+  public Mutation lineage() {
```
Don't extend this class from `Lineage` since it is impossible to emit lineage correctly for this directive, better not support it rather than giving wrong information

yaojiefeng commented on wrangler-core/src/main/java/io/cdap/wrangler/utils/SchemaConverter.java:
```diff
@@ -31,6 +32,7 @@
 import io.cdap.wrangler.api.Row;
 import org.json.JSONException;
 
+
```
nit - remove extra line

yaojiefeng commented on wrangler-core/src/test/java/io/cdap/directives/column/CreateRecordTest.java:
```diff
@@ -0,0 +1,67 @@
+/*
+ *  Copyright © 2017-2019 Cask Data, Inc.
```
should be 2021

rmstar replied:
2021

yaojiefeng commented on wrangler-core/src/test/java/io/cdap/directives/column/ExplodeRecordTest.java:
```diff
@@ -0,0 +1,71 @@
+/*
+ *  Copyright © 2017-2019 Cask Data, Inc.
```
Should be 2021

yaojiefeng commented on wrangler-core/src/test/java/io/cdap/directives/column/ExplodeRecordTest.java:
```diff
@@ -0,0 +1,71 @@
+/**
+ * Tests {@link ExplodeRecordTest}
+ */
+public class ExplodeRecordTest {
```
This test is too simple, should have a test for a more nested scenario.

yaojiefeng commented on wrangler-service/src/main/java/io/cdap/wrangler/service/directive/AbstractDirectiveHandler.java:
```diff
@@ -66,6 +75,8 @@
 import java.util.Set;
 import java.util.stream.Collectors;
 
+
```
nit - remove extra line

yaojiefeng commented on wrangler-service/src/main/java/io/cdap/wrangler/service/directive/AbstractDirectiveHandler.java:
```diff
@@ -17,6 +17,14 @@
 
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.TypeAdapter;
```
these should be unused now right?

yaojiefeng commented on wrangler-service/src/main/java/io/cdap/wrangler/service/directive/AbstractDirectiveHandler.java:
```diff
@@ -146,6 +158,9 @@ public void destroy() {
   }
 
+
+
```
nit - remove extra lines

rmstar replied:
remove extra newlines?

yaojiefeng commented on wrangler-service/src/main/java/io/cdap/wrangler/service/directive/AbstractDirectiveHandler.java:
```diff
@@ -181,11 +198,19 @@ protected DirectiveExecutionResponse generateExecutionResponse(
-          if ((object.getClass().getMethod("toString").getDeclaringClass() != Object.class)) {
-            value.put(fieldName, object.toString());
+
+          if ((object instanceof StructuredRecord)
```
under what condition can this object be `StructuredRecord`? It is not possible that it is still a StructuredRecord right?

yaojiefeng commented on wrangler-service/src/main/java/io/cdap/wrangler/service/directive/WranglerDisplaySerializer.java:
```diff
@@ -0,0 +1,126 @@
+
+
+  private TypeAdapter<Row> rowClassAdapter(Gson gson) {
+    final TypeAdapter<Object> elementAdapter = gson.getAdapter(Object.class);
```
can be private

yaojiefeng commented on wrangler-service/src/main/java/io/cdap/wrangler/service/directive/WranglerDisplaySerializer.java:
```diff
@@ -0,0 +1,126 @@
+        }
+      }
+
+
```
nit - remove extra line

yaojiefeng commented on wrangler-service/src/main/java/io/cdap/wrangler/service/directive/WranglerDisplaySerializer.java:
```diff
@@ -0,0 +1,126 @@
+            out.value(NONDISPLAYABLE_STRING);
+          }
+        } catch (NoSuchMethodException e) {
+          e.printStackTrace();
```
remove this print stack trace

yaojiefeng commented on wrangler-service/src/main/java/io/cdap/wrangler/service/directive/WranglerDisplaySerializer.java:
```diff
@@ -0,0 +1,126 @@
+      return (TypeAdapter<T>) rowClassAdapter(gson);
+    }
+
+    if (Buffer.class.isAssignableFrom(type.getRawType())) {
```
What is the current behavior of dealing Buffer, are we showing the actual bytes value? This change will be backward incompatible.

sechegaray replied:
As mentioned offline. This mirrors current behavior.

yaojiefeng replied:
The original code does not have any special handling of the Bytes type and still show it non-displayable, I don't think we need this check for special handling of Buffer, since all Buffer will be converted to byte[] in https://github.com/data-integrations/wrangler/blob/develop/wrangler-core/src/main/java/io/cdap/wrangler/utils/StructuredToRowTransformer.java#L81-L83. Let's remove this check and add it back when we support other Buffer type in the future.

yaojiefeng commented on wrangler-service/src/main/java/io/cdap/wrangler/service/directive/WranglerDisplaySerializer.java:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    if (Buffer.class.isAssignableFrom(type.getRawType())) {
+      return (TypeAdapter<T>) bufferClassAdapter(gson);
```
This `Buffer` handling and the default object handling can be put together in one method, 
the `if` condition can check whether this is a Buffer or something that doesn't have a `toString` method, this will make the handling clearer

sechegaray replied:
Having as separate methods show a more explicit approach. Having the default check for different classes that we might or not support might make it more complex.
Here we are explicitly mentioning that we are addressing Rows and Buffers explicitly, and everything else gets a toString call (if they have one).  Hiding the functionality in the default method would make it less clear.

yaojiefeng reviewed (commented):
Some more comments

rmstar commented on wrangler-core/src/main/java/io/cdap/directives/column/ExplodeRecord.java:
```diff
@@ -0,0 +1,141 @@
+ * {B: 1, C: 2}
+ *
+ * Directive
+ * exploder-record :A
```
nit: explode-record

bdmogal commented on wrangler-core/src/main/java/io/cdap/directives/column/ExplodeRecord.java:
```diff
@@ -0,0 +1,141 @@
+@Categories(categories = { "column"})
+@Description("Flattens a record into individual columns.")
+public class ExplodeRecord implements Directive, Lineage {
+  public static final String NAME = "explode-record";
```
I know we have an ``explode`` directive. However, can we start using ``flatten-record`` instead, just to make it a less aggressive name?

yaojiefeng commented on wrangler-core/src/main/java/io/cdap/directives/column/FlattenRecord.java:
```diff
@@ -0,0 +1,125 @@
+  @Override
+  public UsageDefinition define() {
+    UsageDefinition.Builder builder = UsageDefinition.builder(NAME);
+    builder.define("column", TokenType.COLUMN_NAME_LIST);
```
call this `columns` if it is for multiple columns

yaojiefeng commented on wrangler-core/src/main/java/io/cdap/wrangler/utils/SchemaConverter.java:
```diff
@@ -45,6 +46,7 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.UUID;
```
nit - unused?

yaojiefeng commented on wrangler-service/src/main/java/io/cdap/wrangler/service/directive/AbstractDirectiveHandler.java:
```diff
@@ -73,11 +76,14 @@
   protected static final String RECORD_DELIMITER_HEADER = "recorddelimiter";
   protected static final String DELIMITER_HEADER = "delimiter";
 
+
```
nit - remove extra line

yaojiefeng commented on wrangler-service/src/main/java/io/cdap/wrangler/service/directive/AbstractDirectiveHandler.java:
```diff
@@ -156,6 +162,8 @@ protected DirectiveExecutionResponse generateExecutionResponse(
     Set<String> headers = new LinkedHashSet<>();
     SchemaConverter convertor = new SchemaConverter();
 
+
```
nit - remove extra line

yaojiefeng reviewed (commented):
Just minor comments

yaojiefeng reviewed (approved):
LGTM, please squash the commits into 1 before merging

sechegaray committed:
Added record support to Wrangler. Converts nested structuredrecords into rows of rows.
Added Json-like display for Structs.
Added two directives (flatten and create) to deal with record types.
